src/context.rs

```diff
@@ -139,26 +138,101 @@ impl<'a> Context<'a> {
             })
     }
 
-    pub fn get_dir_files(&self) -> Result<&Vec<PathBuf>, std::io::Error> {
-        let start_time = SystemTime::now();
-        let scan_timeout = Duration::from_millis(self.config.get_root_config().scan_timeout);
-
-        self.dir_files
-            .get_or_try_init(|| -> Result<Vec<PathBuf>, std::io::Error> {
-                let dir_files = fs::read_dir(&self.current_dir)?
-                    .take_while(|_item| {
-                        SystemTime::now().duration_since(start_time).unwrap() < scan_timeout
-                    })
-                    .filter_map(Result::ok)
-                    .map(|entry| entry.path())
-                    .collect::<Vec<PathBuf>>();
-
-                log::trace!(
-                    "Building a vector of directory files took {:?}",
-                    SystemTime::now().duration_since(start_time).unwrap()
-                );
-                Ok(dir_files)
-            })
+    pub fn dir_contents(&self) -> Result<&DirContents, std::io::Error> {
+        self.dir_contents.get_or_try_init(|| {
+            let timeout = Duration::from_millis(self.config.get_root_config().scan_timeout);
+            DirContents::from_path_with_timeout(&self.current_dir, timeout)
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct DirContents {
+    // HashSet of all files, no folders, relative to the base directory given at construction.
+    files: HashSet<PathBuf>,
+    // HashSet of all file names, e.g. the last section without any folders, as strings.
+    file_names: HashSet<String>,
+    // HashSet of all folders, relative to the base directory given at construction.
+    folders: HashSet<PathBuf>,
+    // HashSet of all extensions found, without dots, e.g. "js" instead of ".js".
+    extensions: HashSet<String>,
+}
+
+impl DirContents {
+    fn from_path(base: &PathBuf) -> Result<Self, std::io::Error> {
```

Comment:
It looks like this function is unused. Perhaps we should remove it?

Reply:
It's used in the tests to avoid having to pass in a bogus Duration. Is that okay?